src/stcal/resample/utils.py

```diff
@@ -0,0 +1,298 @@
+        wcs.pixel_shape = af.tree.get("pixel_shape", None)
+        wcs.array_shape = af.tree.get("array_shape", None)
+
+    if output_shape is not None:
+        wcs.array_shape = output_shape[::-1]
+        wcs.pixel_shape = output_shape
+    elif wcs.pixel_shape is not None:
+        wcs.array_shape = wcs.pixel_shape[::-1]
+    elif wcs.array_shape is not None:
+        wcs.pixel_shape = wcs.array_shape[::-1]
+    elif wcs.bounding_box is not None:
+        wcs.array_shape = tuple(
+            int(axs[1] + 0.5)
+            for axs in wcs.bounding_box.bounding_box(order="C")
+        )
+    else:
+        raise ValueError(
+            "Step argument 'output_shape' is required when custom WCS "
+            "does not have neither of 'array_shape', 'pixel_shape', or "
+            "'bounding_box' attributes set."
+        )
+
+    return wcs
+
+
+def build_mask(dqarr, bitvalue, flag_name_map=None):
+    """Build a bit mask from an input DQ array and a bitvalue flag
+
+    In the returned bit mask, 1 is good, 0 is bad
+    """
+    bitvalue = interpret_bit_flags(bitvalue, flag_name_map=flag_name_map)
+
+    if bitvalue is None:
+        return np.ones(dqarr.shape, dtype=np.uint8)
+    return np.logical_not(np.bitwise_and(dqarr, ~bitvalue)).astype(np.uint8)
+
+
+def get_tmeasure(model):
+    """
+    Check if the measurement_time keyword is present in the datamodel
+    for use in exptime weighting. If not, revert to using exposure_time.
+
+    Returns a tuple of (exptime, is_measurement_time)
+    """
+    try:
+        tmeasure = model["measurement_time"]
+    except KeyError:
+        return model["exposure_time"], False
+    if tmeasure is None:
+        return model["exposure_time"], False
+    else:
+        return tmeasure, True
+
+
+# FIXME: temporarily copied here to avoid this import:
+# from stdatamodels.jwst.library.basic_utils import bytes2human
+def bytes2human(n):
+    """Convert bytes to human-readable format
+
+    Taken from the `psutil` library which references
+    http://code.activestate.com/recipes/578019
+
+    Parameters
+    ----------
+    n : int
+        Number to convert
+
+    Returns
+    -------
+    readable : str
+        A string with units attached.
+
+    Examples
+    --------
+    >>> bytes2human(10000)
+        '9.8K'
+
+    >>> bytes2human(100001221)
+        '95.4M'
+    """
+    symbols = ('K', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y')
+    prefix = {}
+    for i, s in enumerate(symbols):
+        prefix[s] = 1 << (i + 1) * 10
+    for s in reversed(symbols):
+        if n >= prefix[s]:
+            value = float(n) / prefix[s]
+            return '%.1f%s' % (value, s)
+    return "%sB" % n
+
+
+def is_imaging_wcs(wcs):
+    """ Returns `True` if ``wcs`` is an imaging WCS and `False` otherwise. """
+    imaging = all(
+        ax == 'SPATIAL' for ax in wcs.output_frame.axes_type
+    )
+    return imaging
+
+
+def compute_wcs_pixel_area(wcs, shape=None):
+    """ Computes pixel area in steradians.
```

Comment:
The purpose of this PR is to move code that may be shared by multiple pipelines (JWST, Roman, ...) to `stcal`. This PR attempts to preserve how current pipelines work and I do not think this should be modified here.

 > I'm not happy with this and the following routine. It is I think overly complex and doesn't even explain the methodology of how the pixel area is computed.

I will add a little more documentation.

> But why not simply compute the area of one pixel in the center?

We would not need to compute anything if we knew where/how `PIXAR_*` were computed and them maybe we could used them directly without computing anything.

> I gather that it is an average area of pixels over the whole valid image area. But why not simply compute the area of one pixel in the center?

It is too complicated to describe it here but this came from the ambiguity of how `PIXAR_SR` was defined for JWST  images. I attempted to provide a longer explanation here: https://github.com/spacetelescope/jwst/pull/7894#issuecomment-1720338980

Further reading: https://github.com/spacetelescope/jwst/issues/7668#issuecomment-1715808893 and other comments in that issue.

> Perhaps it can be handled in a subsequent PR if this is urgent.

Again, this is code already used in JWST. Please open an issue if you feel this is incorrect.